Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp

```diff
@@ -44,25 +44,64 @@
 namespace WebKit {
 using namespace WebCore;
 
+class RemoteLayerTreeEventDispatcherDisplayLinkClient final : public DisplayLink::Client {
+public:
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    explicit RemoteLayerTreeEventDispatcherDisplayLinkClient(RemoteLayerTreeEventDispatcher& eventDispatcher)
+        : m_eventDispatcher(&eventDispatcher)
+    {
+    }
+    
+    void invalidate()
+    {
+        Locker locker(m_eventDispatcherLock);
+        m_eventDispatcher = nullptr;
+    }
+
+private:
+    // This is called on the display link callback thread.
+    void displayLinkFired(PlatformDisplayID displayID, DisplayUpdate, bool wantsFullSpeedUpdates, bool anyObserverWantsCallback) override
+    {
+        RefPtr<RemoteLayerTreeEventDispatcher> eventDispatcher;
+        {
+            Locker locker(m_eventDispatcherLock);
+            eventDispatcher = m_eventDispatcher;
+        }
+        if (!eventDispatcher)
+            return;
+
+        ScrollingThread::dispatch([dispatcher = Ref { *eventDispatcher }, displayID] {
+            dispatcher->didRefreshDisplay(displayID);
+        });
+    }
+    
+    Lock m_eventDispatcherLock;
+    RefPtr<RemoteLayerTreeEventDispatcher> m_eventDispatcher;
```

Comment:
There should be a way to annotate that this is guarded by `m_eventDispatcherLock` in a way that the compiler will enforce. @heycam knows more I think.

Reply:
Good point, added `WTF_GUARDED_BY_LOCK`